GVRf/Framework/framework/src/main/jni/engine/renderer/render_state.h

```diff
@@ -0,0 +1,713 @@
+        int getStencilRef() const
+        {
+            return render_flags.stencil_ref;
+        }
+
+        void setStencilRef(int ref)
+        {
+            render_flags.stencil_ref = ref;
+            markDirty();
+        }
+
+        int getStencilFunc() const
+        {
+            return render_flags.GLStencilFunc();
+        }
+
+        void setStencilFunc(int func)
+        {
+            render_flags.stencil_func = render_flags.GVRStencilFunc(func);
+            markDirty();
+        }
+
+        int getStencilPass() const
+        {
+            return render_flags.GLStencilOp(render_flags.stencil_pass);
+        }
+
+        void setStencilPass(int glStencilOp)
+        {
+            render_flags.stencil_pass = render_flags.GVRStencilOp(glStencilOp);
+            markDirty();
+        }
+
+        int getStencilFail() const
+        {
+            return render_flags.GLStencilOp(render_flags.stencil_fail);
+        }
+
+        void setStencilFail(int glStencilOp)
+        {
+            render_flags.stencil_fail = render_flags.GVRStencilOp(glStencilOp);
+            markDirty();
+        }
+
+        int getDepthFail() const
+        {
+            return render_flags.GLStencilOp(render_flags.depth_fail);
+        }
+
+        void setDepthFail(int glStencilOp)
+        {
+            render_flags.depth_fail = render_flags.GVRStencilOp(glStencilOp);
+            markDirty();
+        }
+
+        int getSourceBlendFunc() const
+        {
+            return render_flags.GLBlendFunc(render_flags.source_blend);
+        }
+
+        void setSourceBlendFunc(int glBlendFunc)
+        {
+            render_flags.source_blend = render_flags.GVRBlendFunc(glBlendFunc);
+            markDirty();
+        }
+
+        void setDestBlendFunc(int glBlendFunc)
+        {
+            render_flags.dest_blend = render_flags.GVRBlendFunc(glBlendFunc);
+            markDirty();
+        }
+
+        int getDestBlendFunc() const
+        {
+            return render_flags.GLBlendFunc(render_flags.dest_blend);
+        }
+
+        float getOffsetUnits() const
+        {
+            return offset_units;
+        }
+
+        void setOffsetUnits(float units)
+        {
+            if (offset_units != units)
+            {
+                markDirty();
+                offset_units = units;
+            }
+        }
+
+        float getOffsetFactor() const
+        {
+            return offset_factor;
+        }
+
+        void setOffsetFactor(float f)
+        {
+            if (offset_factor!= f)
+            {
+                markDirty();
+                offset_factor = f;
+            }
+        }
+
+        float getSampleCoverage() const
+        {
+            return sample_coverage;
+        }
+
+        void setSampleCoverage(float f)
+        {
+            if (sample_coverage!= f)
+            {
+                markDirty();
+                sample_coverage = f;
+            }
+        }
+
+        uint64_t getRenderFlags() const
```

Comment:
what is intention of this function?

Reply:
This function is for sorting on render modes. It puts all the render flags in a single 64 bit integer. The only things not in there are 3 float values (polygon offset, ..)

Look at MainSceneSorter::mergeByMode to see its usage. It can also be used as a hash code.

Reply:
why last 3 values are not considered for render modes? we use them in setRenderStates(). shouldn't the hash code contain all the values?